omegaconf/grammar_parser.py

```diff
@@ -14,7 +15,8 @@
 )
 
 # Used to cache grammar objects to avoid re-creating them on each call to `parse()`.
-_grammar_cache = None
+# We use a per-thread cache to make it thread-safe.
+_grammar_cache = threading.local()
```

Comment:
Since someone else might also be calling threading.local() to get the exact object, let's call it `thread_local` and the data object inside `grammar_cache`.

Reply:
Also, Maybe we shouldn't store the thread-local-storage object in the global scope as it might get serialized in some scenarios.

It would probably be better to not store the global at all and always access the grammar cache directly from the TLS in the call-site.

Reply:
Wait -- `threading.local()` creates a new storage on each call as far as I can tell:
```python
In [67]: x = threading.local()

In [68]: y = threading.local()

In [69]: x is y
Out[69]: False

In [70]: x.a = 1

In [71]: y.a
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-71-3c3a974596a6> in <module>
----> 1 y.a

AttributeError: '_thread._local' object has no attribute 'a'
```

Reply:
Gotcha. I think that's a bit unconventional but maybe I am wrong. It's been a while since I last used TLS and that was not in Python.